book/src/dev/rfcs/0004-asynchronous-script-verification.md

```diff
@@ -82,18 +97,162 @@ parallelism.
 # Reference-level explanation
 [reference-level-explanation]: #reference-level-explanation
 
-We add a `Request::AwaitUtxo(OutPoint)` and
-`Response::Utxo(transparent::Output)` to the state protocol. As described
-above, the request name is intended to indicate the request's behavior: the
-request does not resolve until the state layer learns of a UTXO described by
-the request.
+## Data structures
+[data-structures]: #data-structures
+
+We add the following request and response to the state protocol:
+```rust
+enum Request::AwaitSpendableUtxo {
+    outpoint: OutPoint,
+    spend_height: Height,
+    spend_restriction: SpendRestriction,
+}
+
+/// A transaction with one or more transparent inputs from coinbase transactions
+/// MUST have no transparent outputs (i.e.tx_out_count MUST be 0).
+enum SpendRestriction {
+    /// The UTXO is spent in a transaction with transparent outputs
+    SomeTransparentOutputs,
+    /// The UTXO is spent in a transaction with all shielded outputs
+    AllShieldedOutputs,
+}
+
+enum Response::Utxo(transparent::Output)
+```
+
+As described above, the request name is intended to indicate the request's behavior.
+The request does not resolve until:
+- the state layer learns of a UTXO described by the request, and
+- the output is spendable at `height` with `spend_restriction`.
+
+We also add a coinbase flag and height to `transparent::Output`s that we look up in
+the state, or get from newly commited blocks:
+```rust
+enum TransparentOutputOrigin {
+    Coinbase {
+        output: transparent::Output,
+        height: Height,
+    },
+    NonCoinbase {
+        output: transparent::Output,
+    },
+}
+```
+
+The coinbase flag and height can be derived from the `tx_index` and `height`:
+- looked up from the `OutPoint.hash` in `tx_by_hash`, or
+- retrieved from the newly arrived block.
+
+There is exactly one coinbase transaction per block, and it must have an index of `0`.
```

Comment:
The existing `Utxo` struct already stores this data, so I need to move the following changes to follow-up tickets:
- refactor into an enum like `TransparentOutputOrigin`: saves disk space
- lookup UTXO details from `OutPoint` rather than duplicating them: spends more CPU and disk accesses, saves disk space

Reply:
This is now #2455.

Reply:
Fixed in f38c87ee